adam_core/coordinates/covariances.py

```diff
@@ -78,12 +78,35 @@ def to_matrix(self) -> np.ndarray:
             Covariance matrices for N coordinates in 6 dimensions.
         """
         # return self.values.combine_chunks().to_numpy_ndarray()
-        cov = np.stack(self.values.to_numpy(zero_copy_only=False))
-        if np.all(cov == None):  # noqa: E711
+        values = self.values.to_numpy(zero_copy_only=False)
+
+        # If all covariance matrices are None, then return a covariances
+        # filled with NaNs.
+        if np.all(values == None):  # noqa: E711
             return np.full((len(self), 6, 6), np.nan)
+
         else:
-            cov = np.stack(cov).reshape(-1, 6, 6)
-        return cov
+            # Try to stack the values into a 3D array. If this works, then
+            # all covariance matrices are the same size and we can return
+            # the stacked matrices.
+            try:
+                cov = np.stack(values).reshape(-1, 6, 6)
+
+            # If not then some of the arrays might be None. Lets loop through
+            # the values and fill in the arrays that are missing (None) with NaNs.
+            except ValueError as e:
+                # If we don't get the error we expect, then raise it.
+                if str(e) != "all input arrays must have the same shape":
+                    raise e
+                else:
+                    for i in range(len(values)):
+                        if values[i] is None:
+                            values[i] = np.full(36, np.nan)
+
+                # Try stacking again
+                cov = np.stack(values).reshape(-1, 6, 6)
+
+            return cov
```

Comment:
I think it's better to do most of this work in Arrow space. Here are a few tools you could use.

1. `self.values.value_lengths()` will give you an array of the lengths of each list element. If a list is null, you'll get a null length:
```py
In [11]: a = pa.array([[1, 2, 3], [4, 5, 6], [7, 8, 9], None])

In [12]: a.value_lengths()
Out[12]:
<pyarrow.lib.Int32Array object at 0x1287c7f40>
[
  3,
  3,
  3,
  null
]
```

2. `Array.value_lengths().unique()` does what you'd hope:
```py
In [13]: a.value_lengths().unique()
Out[13]:
<pyarrow.lib.Int32Array object at 0x1287c7e80>
[
  3,
  null
]
```

3. Instead of iterating over the values to find the nulls, you can efficiently get their indexes with `Array.is_null()`:
```
In [42]: a.is_null()
Out[42]:
<pyarrow.lib.BooleanArray object at 0x12f1df2e0>
[
  false,
  false,
  false,
  true
]
```
You can convert that boolean mask to indices with `pa.compute.indices_nonzero`:
```
In [54]: pa.compute.indices_nonzero(a.is_null())
Out[54]:
<pyarrow.lib.UInt64Array object at 0x1381347c0>
[
  3
]
```

I think you could use these to know very efficiently
1. whether reshaping will work
2. whether it would fail because of nulls vs mis-shapen data
3. which values need to be filled